packages/core/src/input/InputManager.ts

```diff
@@ -0,0 +1,320 @@
+    for (let i = cameras.length - 1; i >= 0; i--) {
+      const camera = cameras[i];
+      if (!camera.enabled || camera.renderTarget) {
+        continue;
+      }
+      const { x: vpX, y: vpY, z: vpW, w: vpH } = camera.viewport;
+      if (x >= vpX && y >= vpY && x - vpX <= vpW && y - vpY <= vpH) {
+        const { _tempHitResult, _tempPoint } = InputManager;
+        _tempPoint.setValue((x - vpX) / vpW, (y - vpY) / vpH);
+        // TODO: Only check which colliders have listened to the input.
+        return engine.physicsManager.raycast(
+          camera.viewportPointToRay(_tempPoint, InputManager._tempRay),
+          _tempHitResult
+        )
+          ? _tempHitResult.collider.entity
+          : null;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Update pointers info.
+   * @param type - Pointer change type
+   * @param pointerId - Pointer ID
+   * @param x - The x coordinate of Pointer
+   * @param y - The y coordinate of Pointer
+   */
+  private _changePointer(type: PointerOperationType, pointerId: number, x?: number, y?: number): void {
+    const {
+      _pointerIdToIndexMap: _pointerIdToIndex,
+      _pointerList: pointerList,
+      _activePointerCount: lastCount,
+      _input: input
+    } = this;
+    let idx = _pointerIdToIndex[pointerId];
+    switch (type) {
+      case PointerOperationType.Remove:
+        if (idx === undefined) {
+          return;
+        } else {
+          const nowCount = (this._activePointerCount = lastCount - 1);
+          if (nowCount != 0) {
+            const removedTouch = pointerList[idx];
+            if (idx !== lastCount) {
+              pointerList[idx] = pointerList[lastCount];
+              pointerList[lastCount] = removedTouch;
+            }
+            input.x = (input.x * lastCount - removedTouch.position.x) / nowCount;
+            input.y = (input.y * lastCount - removedTouch.position.y) / nowCount;
+          }
+          delete _pointerIdToIndex[pointerId];
+        }
+        break;
+      case PointerOperationType.Update:
+        if (idx === undefined) {
+          if (lastCount > 0 && !this._multiTouchEnabled) {
+            return;
+          }
+          let touch: Pointer;
+          if (!pointerList[lastCount]) {
+            touch = pointerList[lastCount] = new Pointer(pointerId, x, y, lastCount);
+          } else {
+            touch = pointerList[lastCount];
+            touch.position.setValue(x, y);
+            touch._indexInList = lastCount;
+          }
+          _pointerIdToIndex[pointerId] = lastCount;
+          const nowCount = (this._activePointerCount = lastCount + 1);
+          input.x = (input.x * lastCount + x) / nowCount;
+          input.y = (input.y * lastCount + y) / nowCount;
+        } else {
+          const { position } = pointerList[idx];
+          input.x += (x - position.x) / lastCount;
+          input.y += (y - position.y) / lastCount;
+          position.setValue(x, y);
+        }
```

Comment:
`if (this._multiTouchEnabled || lastCount < 0)` can get better performance.
```ts
if (this._multiTouchEnabled || lastCount < 0) {
            let touch: Pointer;
            if (!pointerList[lastCount]) {
              touch = pointerList[lastCount] = new Pointer(pointerId, x, y, lastCount);
            } else {
              touch = pointerList[lastCount];
              touch.position.setValue(x, y);
              touch._indexInList = lastCount;
            }
            pointerIdToIndex[pointerId] = lastCount;
            const nowCount = (this._activePointerCount = lastCount + 1);
            input.x = (input.x * lastCount + x) / nowCount;
            input.y = (input.y * lastCount + y) / nowCount;
          } else {
            const { position } = pointerList[index];
            input.x += (x - position.x) / lastCount;
            input.y += (y - position.y) / lastCount;
            position.setValue(x, y);
          }
        }
```

Reply:
Done.